### PR TITLE
PHP 8.1 support

### LINKED ISSUE #8: PHP 8.1 support
### Feature Request

Add support for PHP 8.1

|    Q        |   A
|------------ | ------
| New Feature | yes

### PR DESCRIPTION
**Pull request by MatyCZ**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Initial PHP 8.1 support.

Fixes #8

### CODE REVIEW

File: src/DbSelect.php
Comment (Ocramius) [inline src/DbSelect.php]: Adding type declarations here is a problem, because this class is not `final`, so subtypes are potentially broken.

File: src/ConfigProvider.php
Comment (Ocramius) [inline src/ConfigProvider.php]: Adding type declarations here is a problem, because this class is not `final`, so subtypes are potentially broken.

**Review by Ocramius — CHANGES_REQUESTED** [review]
Please check CI failures too

File: src/DbSelect.php
Comment (MatyCZ) [reply src/DbSelect.php]: Removed in new PR

File: src/ConfigProvider.php
Comment (MatyCZ) [reply src/ConfigProvider.php]: Removed in new PR

**Comment by MatyCZ** [comment]
CI failures are caused by wrong type hints in another libraries (laminas-db), different dependencies (CI test on PHP 7.3 fails due to psr/container needs PHP 7.4, ...) or wrong return type of Mock.

Could you please check CI failures? There is probably nothing to fix.

**Comment by MatyCZ** [comment]
I created another pull request https://github.com/laminas/laminas-db/pull/236. This will fix psalm failure because of wrong type in `Sql\Select::having()`.

File: composer.json
Comment (tobias-trozowski) [inline composer.json]:
```suggestion
        "php": "^7.4 || ~8.0.0 || ~8.1.0",
```
IMO 7.3 support can be droppeed here too as many (most?) other laminas packages already dropped it.

File: composer.json
Comment (tobias-trozowski) [inline composer.json]:
```suggestion
        "phpunit/phpunit": "^9.5.8",
```
bumping phpunit would make it compatible in all combinations (7.4, 8.0, 8.1, lowest, locked, latest)

File: src/DbSelect.php
Comment (froschdesign) [inline src/DbSelect.php]: Please remove all changes like this because the classes are not final and this means a BC break.
Thanks in advance! 👍

File: src/DbSelect.php
Comment (froschdesign) [inline src/DbSelect.php]: The same here.

File: src/DbSelect.php
Comment (froschdesign) [inline src/DbSelect.php]: The same here.

File: src/DbSelect.php
Comment (froschdesign) [inline src/DbSelect.php]: These two lines can be removed because the method signature already gives the same information.

File: src/ConfigProvider.php
Comment (Ocramius) [inline src/ConfigProvider.php]: This class is not `final`, so we can't add return type declarations, as they may break subtypes

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by Ocramius** [comment]
A `composer update` with lowest dependencies (executed on a PHP 7.4 environment) is needed here

**Comment by froschdesign** [comment]
@MatyCZ 
Please check the results of the code sniffer:

```
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$sql has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$select has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$countSelect has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$resultSetPrototype has useless @var annotation.
Error: Property \Laminas\Paginator\Adapter\LaminasDb\DbSelect::$rowCount has useless @var annotation.
```
https://github.com/laminas/laminas-paginator-adapter-laminasdb/pull/9/files#diff-f8dc66f5df7573cebf9d8d594235ad4b1510d2fe3372bf8dd8a05bde4756129dR25

And the Psalm error:
```
Error: src/DbTableGateway.php:41:29: PossiblyInvalidArgument: Argument 1 of Laminas\Db\Sql\Select::having expects Closure|Laminas\Db\Sql\Where|array<array-key, mixed>|string, possibly different type Closure|Laminas\Db\Sql\Having|non-empty-array<array-key, mixed>|non-falsy-string provided (see https://psalm.dev/092)
```
https://github.com/laminas/laminas-paginator-adapter-laminasdb/runs/4583362117?check_suite_focus=true#step:3:289

Thanks in advance! 👍

**Comment by MatyCZ** [comment]
@froschdesign   I am unable to fix psalm error:
https://github.com/laminas/laminas-paginator-adapter-laminasdb/pull/9#issuecomment-996175983

**Comment by weierophinney** [comment]
@MatyCZ I pushed an update to the Psalm baseline to ignore that error; it's a problem with the upstream laminas-db library, so we can ignore it for now.

However... we cannot accept this patch without signoff. You can add it to your existing commits by following [the directions in our contributor document](https://github.com/laminas/.github/blob/main/CONTRIBUTING.md#adding-signoffs-after-the-fact). Once you've done that, ping me, and I'll merge.

**Comment by MatyCZ** [comment]
@weierophinney ping

File: src/DbSelect.php
Comment (froschdesign) [inline src/DbSelect.php]: This is a BC break because the current class is not final and the properties are _only_ protected.

**Review by froschdesign — CHANGES_REQUESTED** [review]

File: src/DbSelect.php
Comment (MatyCZ) [reply src/DbSelect.php]: Removed

File: src/DbSelect.php
Comment (tobias-trozowski) [inline src/DbSelect.php]:
```suggestion
    /** @var Select */
    protected $select;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

File: src/DbSelect.php
Comment (tobias-trozowski) [inline src/DbSelect.php]:
```suggestion
    /** @var Select|null */
    protected $countSelect;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

File: src/DbSelect.php
Comment (tobias-trozowski) [inline src/DbSelect.php]:
```suggestion
    /** @var ResultSetInterface */
    protected $resultSetPrototype;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

File: src/DbSelect.php
Comment (tobias-trozowski) [inline src/DbSelect.php]:
```suggestion
    /** @var null|int*/
    protected $rowCount;
```
As @froschdesign already mentioned
> This is a BC break because the current class is not final and the properties are only protected.

**Review by Ocramius — APPROVED** [review]
LGTM, thanks @MatyCZ!

**Pull request merged**
